File: Session13/90. Subsets II.java
class Solution {
    List<List<Integer>> ans=new ArrayList<>();
public List<List<Integer>> subsetsWithDup(int[] nums) {
Arrays.sort(nums);
subsetsWithDup_bk(nums,new ArrayList<Integer>(),0,nums.length);
return ans;
}

public void subsetsWithDup_bk(int[] nums,List<Integer> temp,int pos,int n)
{
   if(pos==n)
    ans.add(new ArrayList<Integer>(temp));
    return;
}

// Decision 1 taking it 
temp.add(nums[pos]);
subsetsWithDup_bk(nums,temp,pos+1,n);
temp.remove(temp.size()-1);

// Decision 2 not taking it
while(pos+1<n && nums[pos]==nums[pos+1])
{
    pos++;
}
subsetsWithDup_bk(nums,temp,pos+1,n);
}
}

File: Session13/90. Subsets II.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        n=len(nums)
        ans = []
        nums.sort()
        def helper(pos,n,nums,temp):
            if pos>=n:
                ans.append(temp[::])
                return
            temp.append(nums[pos])
            helper(pos+1,n,nums,temp)
            temp.pop(-1)
            while pos<n-1 and nums[pos]==nums[pos+1]:
                pos+=1
            helper(pos+1,n,nums,temp)
        helper(0,n,nums,[])
        return ans
            